worker/caasapplicationprovisioner/application.go

```diff
@@ -0,0 +1,311 @@
+			Site: &a.catacomb,
+			Work: a.loop,
+		})
+		return a, err
+	}
+}
+
+func (a *appWorker) Notify() {
+	a.changes <- struct{}{}
+}
+
+func (a *appWorker) Kill() {
+	a.catacomb.Kill(nil)
+}
+
+func (a *appWorker) Wait() error {
+	return a.catacomb.Wait()
+}
+
+func (a *appWorker) loop() error {
+	var err error
+	var appLife life.Value = life.Dead
+
+	charmURL, err := a.facade.ApplicationCharmURL(a.name)
+	if err != nil {
+		return errors.Annotatef(err, "failed to get charm urls for application")
+	}
+	charmInfo, err := a.facade.CharmInfo(charmURL.String())
+	if err != nil {
+		return errors.Annotatef(err, "failed to get application charm deployment metadata for %q", a.name)
+	}
+	if charmInfo == nil ||
+		charmInfo.Meta == nil ||
+		charmInfo.Meta.Deployment == nil ||
+		charmInfo.Meta.Deployment.DeploymentMode != charm.ModeEmbedded {
+		a.logger.Debugf("skipping non-embedded application %q", a.name)
+		a.catacomb.Kill(nil)
+		return nil
+	}
+
+	app := a.broker.Application(a.name,
+		caas.DeploymentType(charmInfo.Meta.Deployment.DeploymentType))
+
+	var appChanges watcher.NotifyChannel
+	var replicaChanges watcher.NotifyChannel
+
+	for {
+		select {
+		case <-a.catacomb.Dead():
+			return a.catacomb.Err()
+		case <-a.changes:
+			appLife, err = a.facade.Life(a.name)
+			if errors.IsNotFound(err) {
+				appLife = life.Dead
+			} else if err != nil {
+				return errors.Trace(err)
+			}
+			switch appLife {
+			case life.Alive:
+				err = a.alive(app)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				if appChanges == nil {
+					appWatcher, err := app.Watch()
+					if err != nil {
+						return errors.Annotatef(err, "failed to watch for changes to application %q", a.name)
+					}
+					a.catacomb.Add(appWatcher)
+					if err != nil {
+						appWatcher.Kill()
+						return errors.Trace(err)
+					}
+					appChanges = appWatcher.Changes()
+				}
+				if replicaChanges == nil {
+					replicaWatcher, err := app.WatchReplicas()
+					if err != nil {
+						return errors.Annotatef(err, "failed to watch for changes to replicas %q", a.name)
+					}
+					a.catacomb.Add(replicaWatcher)
+					if err != nil {
+						replicaWatcher.Kill()
+						return errors.Trace(err)
+					}
+					replicaChanges = replicaWatcher.Changes()
+				}
+			case life.Dying:
+				err = a.dying(app)
+				if err != nil {
+					return errors.Trace(err)
+				}
+			case life.Dead:
+				return a.dead(app)
+			}
+		case <-appChanges:
+			err = a.updateState(app, false)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		case <-replicaChanges:
+			err = a.updateState(app, false)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+}
+
+func (a *appWorker) updateState(app caas.Application, force bool) error {
+	// Fetching the units here is to ensure happens-before consistency
+	// on the deletion of units.
+	observedUnits, err := a.facade.Units(a.name)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	st, err := app.State()
+	if errors.IsNotFound(err) {
```

Comment:
return nil ?

Reply:
Returning nil here would prevent the call to force cleanup remaining units.